potential overflow in use of strcpy

Hello! Thank you for devilutionx!

I bought a bow from Griswold, it needs 70 dexterity, which are available. But it cannot be equipped.

First image shows the requirement
![screen00](https://user-images.githubusercontent.com/1104628/97804396-f0a0db80-1c4f-11eb-8dca-2a908d5462cf.png)

I tried undressing all items and am holding the bow with the mouse and try to equip it, one cannot see that. But the status box indicates it.
![screen01](https://user-images.githubusercontent.com/1104628/97804398-f1397200-1c4f-11eb-8a9e-0f5d630da540.png)

To reproduce, here is a save game. The bow in question lies on the ground next to you.
[single_1.sv.zip](https://github.com/diasurgical/devilutionX/files/5470937/single_1.sv.zip)

Regards

I've found a warning at compile time that I think it's related to this issue
```
In function ‘strcpy’,
    inlined from ‘AddPanelString’ at /home/gianluca/devilutionx/src/devilutionX-1.1.0/Source/control.cpp:552:8,
    inlined from ‘PrintItemDur’ at /home/gianluca/devilutionx/src/devilutionX-1.1.0/Source/items.cpp:3106:18,
    inlined from ‘CheckInvHLight’ at /home/gianluca/devilutionx/src/devilutionX-1.1.0/Source/inv.cpp:2026:16:
/usr/include/bits/string_fortified.h:90:33: warning: ‘__builtin_memcpy’ writing 15 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
   90 |   return __builtin___strcpy_chk (__dest, __src, __bos (__dest));
```

Found the guilty.
_FORTIFY_SOURCE directive (a default CXX Flag for many distros) generate this warning at compile time:
```
[100%] Linking CXX executable devilutionx
In function ‘strcpy’,
    inlined from ‘AddPanelString’ at /home/gianluca/devilutionx/src/devilutionX-1.1.0/Source/control.cpp:552:8,
    inlined from ‘PrintItemDur’ at /home/gianluca/devilutionx/src/devilutionX-1.1.0/Source/items.cpp:3106:18,
    inlined from ‘CheckInvHLight’ at /home/gianluca/devilutionx/src/devilutionX-1.1.0/Source/inv.cpp:2026:16:
/usr/include/bits/string_fortified.h:90:33: warning: ‘__builtin_memcpy’ writing 15 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
   90 |   return __builtin___strcpy_chk (__dest, __src, __bos (__dest));
      |                                 ^
[100%] Built target devilutionx
```
It generates a valid build but as @Tmkrth stated requirements are not correctly printed.

if could be of interest another warning with standard arch linux flags is this reported below
```
[ 82%] Building CXX object CMakeFiles/devilutionx.dir/SourceX/DiabloUI/art_draw.cpp.o
/home/gianluca/devilutionx/src/devilutionX-1.1.0/SourceX/storm/storm_net.cpp: In function ‘dvl::BOOL dvl::SNetUnregisterEventHandler(int, dvl::SEVTHANDLER)’:
/home/gianluca/devilutionx/src/devilutionX-1.1.0/SourceX/storm/storm_net.cpp:53:50: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   53 |  return dvlnet_inst->SNetUnregisterEventHandler(*(event_type *)&evtype, func);
      |                                                  ^~~~~~~~~~~~~~~~~~~~~
/home/gianluca/devilutionx/src/devilutionX-1.1.0/SourceX/storm/storm_net.cpp: In function ‘dvl::BOOL dvl::SNetRegisterEventHandler(int, dvl::SEVTHANDLER)’:
/home/gianluca/devilutionx/src/devilutionX-1.1.0/SourceX/storm/storm_net.cpp:58:48: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   58 |  return dvlnet_inst->SNetRegisterEventHandler(*(event_type *)&evtype, func);
      |                                                ^~~~~~~~~~~~~~~~~~~~~
[ 83%] Building CXX object CMakeFiles/devilutionx.dir/SourceX/DiabloUI/errorart.cpp.o
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -640,7 +640,7 @@ if(NOT CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
   target_compile_options(devilution PUBLIC $<${DEBUG_GENEX}:-fno-omit-frame-pointer>)
 
   # Warnings for devilution
-  target_compile_options(devilution PRIVATE -Wall -Wno-multichar -Wno-int-to-pointer-cast)
+  target_compile_options(devilution PRIVATE -Wno-multichar -Wno-int-to-pointer-cast)
 
   # Warnings for devilutionX
   target_compile_options(${BIN_TARGET} PRIVATE -Wall -Wextra -Wno-unused-parameter -Wno-format-security)

--- a/Source/items.cpp
+++ b/Source/items.cpp
@@ -4160,11 +4160,27 @@ void PrintItemMisc(ItemStruct *x)
 	}
 }
 
+static void PrintItemInfo(ItemStruct *x)
+{
+	PrintItemMisc(x);
+	Uint8 str = x->_iMinStr;
+	Uint8 dex = x->_iMinDex;
+	Uint8 mag = x->_iMinMag;
+	if (str || mag || dex) {
+		strcpy(tempstr, "Required:");
+		if (str)
+			sprintf(tempstr + strlen(tempstr), " %i Str", str);
+		if (mag)
+			sprintf(tempstr + strlen(tempstr), " %i Mag", mag);
+		if (dex)
+			sprintf(tempstr + strlen(tempstr), " %i Dex", dex);
+		AddPanelString(tempstr, TRUE);
+	}
+	pinfoflag = TRUE;
+}
+
 void PrintItemDetails(ItemStruct *x)
 {
-	char str, dex;
-	BYTE mag;
-
 	if (x->_iClass == ICLASS_WEAPON) {
 		if (x->_iMinDam == x->_iMaxDam) {
 			if (x->_iMaxDur == DUR_INDESTRUCTIBLE)
@@ -4207,27 +4223,11 @@ void PrintItemDetails(ItemStruct *x)
 		uitemflag = TRUE;
 		curruitem = *x;
 	}
-	PrintItemMisc(x);
-	mag = x->_iMinMag;
-	dex = x->_iMinDex;
-	str = x->_iMinStr;
-	if (mag + dex + str) {
-		if (x->_iMinStr)
-			sprintf(tempstr, "Required: %i Str", x->_iMinStr);
-		if (x->_iMinMag)
-			sprintf(tempstr, "Required: %i Mag", x->_iMinMag);
-		if (x->_iMinDex)
-			sprintf(tempstr, "Required: %i Dex", x->_iMinDex);
-		AddPanelString(tempstr, TRUE);
-	}
-	pinfoflag = TRUE;
+	PrintItemInfo(x);
 }
 
 void PrintItemDur(ItemStruct *x)
 {
-	char str, dex;
-	BYTE mag;
-
 	if (x->_iClass == ICLASS_WEAPON) {
 		if (x->_iMinDam == x->_iMaxDam) {
 			if (x->_iMaxDur == DUR_INDESTRUCTIBLE)
@@ -4263,20 +4263,7 @@ void PrintItemDur(ItemStruct *x)
 	}
 	if (x->_itype == ITYPE_RING || x->_itype == ITYPE_AMULET)
 		AddPanelString("Not Identified", TRUE);
-	PrintItemMisc(x);
-	str = x->_iMinStr;
-	mag = x->_iMinMag;
-	dex = x->_iMinDex;
-	if (str + mag + dex) {
-		if (x->_iMinStr)
-			sprintf(tempstr, "Required: %i Str", x->_iMinStr);
-		if (x->_iMinMag)
-			sprintf(tempstr, "Required: %i Mag", x->_iMinMag);
-		if (x->_iMinDex)
-			sprintf(tempstr, "Required: %i Dex", x->_iMinDex);
-		AddPanelString(tempstr, TRUE);
-	}
-	pinfoflag = TRUE;
+	PrintItemInfo(x);
 }
 
 void UseItem(int p, int Mid, int spl)

